doc/articles/Ownership.md

```diff
@@ -24,6 +24,14 @@ Note that such functions will *not* free any previously-allocated data: overwrit
 
 Note also that some structs with `FreeMembers` functions may be used as both inputs and outputs. In this case `FreeMembers` must only be used if the member allocations were made by the `webgpu.h` implementation (as an output), not if they were made by the application (as an input).
 
+## Releasing a Device object {#DeviceRelease}
+
+Unlike other destroyable objects, releasing the last (external) ref to a device causes it to be automatically destroyed (via @ref wgpuDeviceDestroy) if it isn't already lost or destroyed. Though the device object is no longer valid to use after releasing the last ref, this has some notable direct effects:
+
+- It destroy buffers, unmapping them, aborting any pending map requests, and preventing future ones.
+- It fires the DeviceLost event (which will call the registered callback at some point, depending on its @ref WGPUCallbackMode).
+    - Because the device's last ref has already been released, DeviceLost callbacks triggered in this way will *not* receive a pointer to the device object. This is still true if the callback was triggered *during* the release of the last ref (via @ref WGPUCallbackMode_AllowSpontaneous).
```

Comment:
This last bit is a bit funky because it means it doesn't literally behave the same way as calling `wgpuDeviceDestroy` (which, since it's called on a device, means the device must necessarily still have refs). I would choose one of the following, in order of preference:

- Lose the device instead of destroying it. This would have the additional effect not unmapping already-mapped buffers.
- Change this bit this so that it will be non-null if called *during* the Release call.
- Describe this as destroying the device but without literally saying it calls `wgpuDeviceDestroy()`. Handwavy.

Reply:
When you say lose the device, would that be a case where objects end up keeping the device alive? Like just the device lost callback gets fired, but everything else just keeps continue working?

Reply:
No, I mean it would be like a "natural" device loss, like a backend-initiated device loss or unexpected OOM. The device is still lost and nothing works, the only difference in the operation itself is that it doesn't destroy buffers.

I poorly explained why that's relevant here. The behavior I want, for consistency, is for Release to do [drop last ref] [trigger destroy or loss], in that order. The problem with that is that you _can't_ (at the API level) trigger destroy without a ref. Whereas device loss can happen at any time.

Reply:
Sorry, I may have forgotten already if we talked about it, but why can't dropping the last ref just look like it calls `wgpuDeviceDestory` again? It doesn't have to be handwavy because it could just do the same things? The only caveat would be the difference in the device lost callback right?

Reply:
That's essentially what the current revision of the PR does.

Since Destroy can't be called without a ref, it's implied that Release looks like:
`if (externalRefCount == 1) { Destroy(); externalRefCount--; }`
- If the DeviceLost callback is non-spontaneous, the callback will get called some time later:
  `Release() -> Destroy()`
  `ProcessEvents()/WaitAny() -> callback`
  When the callback is called, externalRefCount==0.
- If it *is* spontaneous, the callback will get called during Release:
  `Release() -> Destroy() -> callback`
  When the callback is called, externalRefCount==1. Hence the special case in the PR to say this still passes null to the DeviceLost callback, even though it still technically has a ref.

Reply:
Ok, yea that sgtm then!

Reply:
It's kinda weird though. The implementation would be one of the following:
- Take internal ref; free external ref; if externalRefCount==0, call Destroy(); free internal ref.
  (And when calling the DeviceLost callback, we have to choose whether to pass null based on the externalRefCount of the device, which is also weird)
- Internally use a different signal than Destroy(), so that we know to pass null regardless of any refs on the device.

Is it better to do one of the alternatives I mentioned above?

Reply:
> * Internally use a different signal than Destroy(), so that we know to pass null regardless of any refs on the device.

Actually this makes me realize there's another weird case: the device is lost naturally (the future becomes Ready), then the last external ref is dropped, then it calls the DeviceLost callback.
- Do we pass null in this case too? (How do we know to do that? I think the only way is by looking at the external refcount.)
- Or does DeviceLost becoming Ready hold an external ref to the device so that this can't happen?